google-cloud-datastore/lib/google/cloud/datastore/filter.rb

```diff
@@ -0,0 +1,208 @@
+        #
+        def initialize name, operator, value
+          @grpc = create_property_filter name, operator, value
+        end
+
+        ##
+        # Joins two filters with an AND operator.
+        #
+        # @overload and(name, operator, value)
+        #   Joins the filter with a property filter
+        #   @param name [String]
+        #   @param operator [String]
+        #   @param value
+        #
+        # @overload and(filter)
+        #   Joins the filter with a Filter object
+        #   @param flter [Filter]
+        #
+        # @example Join the filter with a property filter
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   filter = Google::Cloud::Filter.new("done", "=", false)
+        #                                 .and("priority", ">=", 4)
+        #
+        # @example Join the filter with a filter object
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   filter_1 = Google::Cloud::Filter.new("done", "=", false)
+        #   filter_2 = Google::Cloud::Filter.new("priority", ">=", 4)
+        #
+        #   filter = filter_1.and(filter_2)
+        #
+        def and name_or_filter, operator = nil, value = nil
+          combine_filters composite_filter_and, name_or_filter, operator, value
+        end
+
+        ##
+        # Joins two filters with an OR operator.
+        #
+        # @overload or(name, operator, value)
+        #   Joins the filter with a property filter
+        #   @param name [String]
+        #   @param operator [String]
+        #   @param value
+        #
+        # @overload or(filter)
+        #   Joins the filter with a Filter object
+        #   @param flter [Filter]
+        #
+        # @example Join the filter with a property filter
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   filter = Google::Cloud::Filter.new("done", "=", false)
+        #                                 .or("priority", ">=", 4)
+        #
+        # @example Join the filter with a filter object
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   filter_1 = Google::Cloud::Filter.new("done", "=", false)
+        #   filter_2 = Google::Cloud::Filter.new("priority", ">=", 4)
+        #
+        #   filter = filter_1.or(filter_2)
+        #
+        def or name_or_filter, operator = nil, value = nil
+          combine_filters composite_filter_or, name_or_filter, operator, value
+        end
+
+        # @private
+        def to_grpc
+          @grpc
+        end
+
+        protected
+
+        ##
+        # @private
+        def combine_filters composite_filter, name_or_filter, operator, value
+          composite_filter.composite_filter.filters << to_grpc
```

Comment:
Can we please rename `composite_filter` to something else to avoid `composite_filter.composite_filter` ?

Reply:
I renamed the parameter from `composite_filter` to `new_filter` to make the intent clearer.